src/eval/context.rs

```diff
@@ -0,0 +1,98 @@
+use std::rc::Rc;
+use std::cell::RefCell;
+use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut};
+
+use super::{Env, eval};
+use ::{Value, AresResult, AresError, parse, stdlib, Environment};
+
+pub struct Context<T> {
+    env: Env,
+    _state: PhantomData<T>,
+}
+
+pub struct LoadedContext<'a, 'b, T: 'a + 'b> {
+    ctx: &'b mut Context<T>,
+    state: &'a mut T,
+}
+
+impl <T> Context<T> {
+    pub fn new() -> Context<T> {
+        let env = Rc::new(RefCell::new(Environment::new()));
+        stdlib::load_all(&env);
+        Context {
+            env: env,
+            _state: PhantomData,
+        }
+    }
+
+    pub fn new_empty() -> Context<T> {
+        Context {
+            env: Rc::new(RefCell::new(Environment::new())),
+            _state: PhantomData,
+        }
+    }
+
+    pub fn load<'a, 'b>(&'b mut self, state: &'a mut T) -> LoadedContext<'a, 'b, T> {
+        LoadedContext {
+            ctx: self,
+            state: state
+        }
+    }
+
+    pub fn env(&self) -> &Env {
+        &self.env
+    }
+}
+
+impl <'a, 'b,  T> LoadedContext<'a, 'b, T> {
+    pub fn eval(&mut self, program: &Value) -> AresResult<Value> {
+        eval(program, self.env())
+    }
+
+    pub fn eval_str(&mut self, program: &str) -> AresResult<Value> {
+        let trees = try!(parse(program));
+        let mut last = None;
+        for tree in trees {
+            last = Some(try!(eval(&tree, self.env())))
+        }
+        match last {
+            Some(v) => Ok(v),
+            None => Err(AresError::NoProgram)
+        }
+    }
+
+    pub fn call(&mut self, func: Value, mut args: Vec<Value>) -> AresResult<Value> {
+        for arg in args.iter_mut() {
+            if let a@&mut Value::List(_) = arg {
+                // TODO: can you get rid of this clone?
+                *a = Value::new_list(vec![Value::new_ident("quote"), a.clone()]);
+            }
+        }
+
+        let built = Value::new_list(vec![func, Value::new_list(args)]);
+        self.eval(&built)
```

Comment:
Well, it would be generally useful too, like here: https://github.com/TyOverby/ares/blob/master/src/stdlib/list.rs#L79-L83

And I was thinking that such a function could actually replace a lot of this code: https://github.com/TyOverby/ares/blob/master/src/eval/mod.rs#L38-L60

basically, evaluate the head of the list, and if it's a procedure or freefn, evaluate the arguments; then call `apply` (or whatever) to do the work of applying the lisp function to the values. In cases like this where we're constructing the function call in Rust, just call apply directly. But this is orthogonal to this PR really---I'll try to work something up after work.